File: packages/webrtc-core/src/sharer-session.ts
import type { ConnectionState, JoinRequestInfo } from '@crossscreen/protocol';

import type { ApiClient, CreatedSession } from './api-client.ts';
import { Emitter } from './events.ts';
import { IceCandidateQueue } from './ice-queue.ts';
import { SignalingClient } from './signaling-client.ts';
import { formatSnapshot, readConnectionSnapshot, type ConnectionSnapshot } from './stats.ts';
import { tuneScreenShare } from './tuning.ts';

/**
 * Sharing a screen, end to end, without knowing what is drawing the screen.
 *
 * Create a session, attach as host, wait for someone to ask, and — only once
 * the host has said yes — negotiate with them. That order is the product's
 * security model (ADR-0006), so it lives here rather than being reassembled
 * correctly by every interface that uses it.
 *
 * Framework-agnostic and event-driven, because the same flow has to serve a
 * browser tab, a React app, and Electron's renderer.
 */

export interface SharerEvents {
  /** Session created; the code and link can be shown. */
  ready: CreatedSession;
  /** Someone is asking to watch. Nothing has been sent to them. */
  pending: JoinRequestInfo;
  viewerJoined: { participantId: string };
  viewerLeft: { participantId: string };
  connection: { state: ConnectionState };
  stats: ConnectionSnapshot;
  /** The share ended — by us, by the OS, or by the far end. */
  ended: { reason: string };
  error: { message: string };
}

export interface SharerDependencies {
  api: ApiClient;
  signalingUrl: string;
  /** Supplied rather than captured here: the shell owns the capture policy. */
  stream: MediaStream;
}

/**
 * Translate an ICE state into words a person can read (architecture §67).
 * `RTCPeerConnectionState` is a debugging vocabulary; nobody should see
 * "failed" and be expected to know what to do about it.
 */
export function userFacingState(state: RTCPeerConnectionState): ConnectionState {
  switch (state) {
    case 'new':
      return 'connecting';
    case 'connecting':
      return 'checking';
    case 'connected':
      return 'connected';
    case 'disconnected':
      return 'unstable';
    case 'failed':
    case 'closed':
      return 'failed';
  }
}

/** Connection quality as the user sees it, from what was measured. */
export function qualityFrom(
  snapshot: ConnectionSnapshot,
): 'excellent' | 'good' | 'poor' | 'unstable' {
  const rtt = snapshot.roundTripMs;
  if (rtt === undefined) return 'good';
  if (rtt < 60) return 'excellent';
  if (rtt < 150) return 'good';
  if (rtt < 400) return 'poor';
  return 'unstable';
}

export class SharerSession extends Emitter<SharerEvents> {
  /** Everyone waiting on the host right now. */
  readonly pending = new Map<string, JoinRequestInfo>();

  #signaling: SignalingClient | undefined;
  readonly #peers = new Map<string, { pc: RTCPeerConnection; ice: IceCandidateQueue }>();
  #iceServers: RTCIceServer[] = [];
  #statsTimer: ReturnType<typeof setInterval> | undefined;
  #session: CreatedSession | undefined;
  #stopped = false;

  readonly #deps: SharerDependencies;

  constructor(deps: SharerDependencies) {
    super();
    this.#deps = deps;
  }

  get session(): CreatedSession | undefined {
    return this.#session;
  }

  async start(): Promise<CreatedSession> {
    // ICE configuration is fetched first so a failure surfaces as "could not
    // start" rather than as a connection that never completes for reasons
    // nobody can see.
    this.#iceServers = await this.#deps.api.iceServers();

    const session = await this.#deps.api.createSession();
    this.#session = session;

    const signaling = new SignalingClient(this.#deps.signalingUrl);
    this.#signaling = signaling;
    this.#wire(signaling);

    await signaling.connect();
    signaling.send({ type: 'session.host.attach', hostToken: session.hostToken });

    // The OS or the browser can end a capture without asking. Treat the track
    // ending as authoritative rather than assuming we are still sharing.
    this.#deps.stream.getVideoTracks()[0]?.addEventListener('ended', () => {
      this.emit('ended', { reason: 'Your device stopped the screen share' });
      this.stop();
    });

    this.emit('ready', session);
    return session;
  }

  /** Let a waiting viewer in. Until this, nothing has been sent to them. */
  approve(participantId: string): void {
    this.pending.delete(participantId);
    this.#signaling?.send({ type: 'session.viewer.approve', participantId });
  }

  reject(participantId: string): void {
    this.pending.delete(participantId);
    this.#signaling?.send({ type: 'session.viewer.reject', participantId });
  }

  stop(): void {
    if (this.#stopped) return;
    this.#stopped = true;

    if (this.#statsTimer !== undefined) clearInterval(this.#statsTimer);
    this.#statsTimer = undefined;

    for (const { pc } of this.#peers.values()) pc.close();
    this.#peers.clear();
    this.pending.clear();

    this.#signaling?.send({ type: 'session.end' });
    this.#signaling?.close();
    this.#signaling = undefined;
  }

  #wire(signaling: SignalingClient): void {
    signaling.on('session.viewer.pending', (message) => {
      // Held, not acted on. The host decides, and until they do, this viewer
      // receives nothing at all.
      this.pending.set(message.request.participantId, message.request);
      this.emit('pending', message.request);
    });

    // Only ever sent after approval, which is what makes offering here safe.
    signaling.on('peer.joined', (message) => {
      void this.#offerTo(message.participant.participantId);
      this.emit('viewerJoined', { participantId: message.participant.participantId });
    });

    signaling.on('peer.left', (message) => {
      this.#closePeer(message.participantId);
      this.emit('viewerLeft', { participantId: message.participantId });
    });

    signaling.on('rtc.answer', async (message) => {
      const peer = this.#peers.get(message.from);
      if (peer === undefined) return;
      await peer.pc.setRemoteDescription({ type: 'answer', sdp: message.sdp });
      await peer.ice.flush();
    });

    signaling.on('rtc.ice', async (message) => {
      await this.#peers.get(message.from)?.ice.add({
        candidate: message.candidate,
        sdpMid: message.sdpMid,
        sdpMLineIndex: message.sdpMLineIndex,
      });
    });

    signaling.on('session.ended', (message) => {
      this.emit('ended', { reason: message.reason });
      this.stop();
    });

    signaling.on('error', (message) => {
      this.emit('error', { message: message.userMessage });
    });
  }

  async #offerTo(participantId: string): Promise<void> {
    const signaling = this.#signaling;
    if (signaling === undefined) return;

    const pc = new RTCPeerConnection({ iceServers: this.#iceServers });
    const ice = new IceCandidateQueue(pc);
    this.#peers.set(participantId, { pc, ice });

    pc.onicecandidate = (event): void => {
      if (event.candidate === null) return;
      signaling.send({
        type: 'rtc.ice',
        to: participantId,
        candidate: event.candidate.candidate,
        sdpMid: event.candidate.sdpMid,
        sdpMLineIndex: event.candidate.sdpMLineIndex,
      });
    };

    pc.onconnectionstatechange = (): void => {
      this.emit('connection', { state: userFacingState(pc.connectionState) });
    };

    const track = this.#deps.stream.getVideoTracks()[0];
    if (track === undefined) {
      this.emit('error', { message: 'The screen share ended before it could be sent' });
      return;
    }

    const sender = pc.addTrack(track, this.#deps.stream);
    const transceiver = pc.getTransceivers().find((t) => t.sender === sender);
    await tuneScreenShare(track, sender, transceiver);

    const offer = await pc.createOffer();
    await pc.setLocalDescription(offer);
    if (offer.sdp === undefined) {
      this.emit('error', { message: 'Could not start the connection' });
      return;
    }
    signaling.send({ type: 'rtc.offer', to: participantId, sdp: offer.sdp });

    this.#startStats();
  }

  #closePeer(participantId: string): void {
    const peer = this.#peers.get(participantId);
    if (peer === undefined) return;
    peer.pc.close();
    this.#peers.delete(participantId);
  }

  #startStats(intervalMs = 2000): void {
    if (this.#statsTimer !== undefined) return;
    this.#statsTimer = setInterval(() => {
      const [first] = this.#peers.values();
      if (first === undefined) return;

      void readConnectionSnapshot(first.pc).then((snapshot) => {
        this.emit('stats', snapshot);
        // Reported so Phase 2 can answer the question that predicts TURN
        // cost: what fraction of connections go direct rather than relayed.
        this.#signaling?.send({
          type: 'stats.report',
          quality: qualityFrom(snapshot),
          connectionState: userFacingState(first.pc.connectionState),
          transport: snapshot.transport,
          ...(snapshot.roundTripMs === undefined ? {} : { roundTripMs: snapshot.roundTripMs }),
          ...(snapshot.resolution === undefined ? {} : { resolution: snapshot.resolution }),
          ...(snapshot.codec === undefined ? {} : { codec: snapshot.codec }),
          ...(snapshot.framesPerSecond === undefined
            ? {}
            : { framesPerSecond: snapshot.framesPerSecond }),
        });
        console.debug('[sharer]', formatSnapshot(snapshot));
      });
    }, intervalMs);
  }
}

File: apps/android/app/src/main/kotlin/app/crossscreen/android/net/SharerSession.kt
package app.crossscreen.android.net

import android.util.Log
import app.crossscreen.android.protocol.ClientMessage
import app.crossscreen.android.protocol.ConnectionState
import app.crossscreen.android.protocol.JoinRequestInfo
import app.crossscreen.android.protocol.ServerMessage
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import org.webrtc.IceCandidate
import org.webrtc.PeerConnection
import org.webrtc.SessionDescription
import org.webrtc.VideoTrack

/**
 * Sharing a phone screen, end to end — the Kotlin counterpart of
 * `packages/webrtc-core/src/sharer-session.ts`, cut to the happy path:
 * create a session, attach as host, wait for someone to ask, and — only
 * once the host says yes — negotiate with them. That order is ADR-0006.
 *
 * Deliberately not ported yet: stats reporting, quality tuning, forced
 * relay, ICE-restart recovery, and more than one viewer at a time. Each is
 * additive on top of a connection that works.
 */
class SharerSession(
    private val scope: CoroutineScope,
    serverBaseUrl: String,
    private val track: VideoTrack,
) {
    data class Ui(
        val joinCode: String? = null,
        val joinCodeDisplay: String? = null,
        val shareLink: String? = null,
        val pending: List<JoinRequestInfo> = emptyList(),
        val viewerCount: Int = 0,
        val connection: ConnectionState = ConnectionState.CONNECTING,
        val error: String? = null,
        val ended: String? = null,
    )

    private val api = ApiClient(serverBaseUrl.trim().trimEnd('/'))
    private val wsUrl = signalingUrlFrom(serverBaseUrl)

    private val _ui = MutableStateFlow(Ui())
    val ui: StateFlow<Ui> = _ui.asStateFlow()

    private var signaling: SignalingClient? = null
    private var hostToken: String? = null
    private var iceServers: List<PeerConnection.IceServer> = emptyList()
    private val peers = mutableMapOf<String, Peer>()
    private var stopped = false

    private class Peer(val pc: PeerConnection, val ice: IceCandidateQueue)

    suspend fun start() {
        iceServers = try {
            api.iceServers()
        } catch (e: ApiException) {
            _ui.update { it.copy(error = e.message) }
            return
        }

        val session = try {
            api.createSession()
        } catch (e: ApiException) {
            _ui.update { it.copy(error = e.message) }
            return
        }
        hostToken = session.hostToken
        _ui.update {
            it.copy(
                joinCode = session.joinCode,
                joinCodeDisplay = session.joinCodeDisplay,
                shareLink = session.shareLink,
            )
        }

        val client = SignalingClient(wsUrl)
        signaling = client
        wire(client)
        try {
            client.connect()
        } catch (e: Exception) {
            _ui.update { it.copy(error = "Couldn't reach the server at $wsUrl.") }
            return
        }
        if (stopped) {
            client.close()
            return
        }
        client.send(ClientMessage.SessionHostAttach(session.hostToken))
    }

    fun approve(participantId: String) {
        _ui.update { it.copy(pending = it.pending.filterNot { p -> p.participantId == participantId }) }
        signaling?.send(ClientMessage.SessionViewerApprove(participantId))
    }

    fun reject(participantId: String) {
        _ui.update { it.copy(pending = it.pending.filterNot { p -> p.participantId == participantId }) }
        signaling?.send(ClientMessage.SessionViewerReject(participantId))
    }

    fun stop() {
        if (stopped) return
        stopped = true
        signaling?.send(ClientMessage.SessionEnd)
        peers.values.forEach { it.pc.dispose() }
        peers.clear()
        signaling?.close()
        signaling = null
    }

    private fun wire(client: SignalingClient) {
        client.on<ServerMessage.SessionViewerPending> { m ->
            _ui.update { it.copy(pending = it.pending + m.request) }
        }
        client.on<ServerMessage.PeerJoined> { m ->
            scope.launch { offerTo(m.participant.participantId) }
        }
        client.on<ServerMessage.RtcAnswer> { m ->
            val peer = peers[m.from] ?: return@on
            scope.launch {
                try {
                    peer.pc.setRemoteAwait(SessionDescription(SessionDescription.Type.ANSWER, m.sdp))
                    peer.ice.onRemoteDescriptionSet()
                } catch (e: RtcException) {
                    Log.w(TAG, "setRemote(answer) failed", e)
                }
            }
        }
        client.on<ServerMessage.RtcIce> { m ->
            peers[m.from]?.ice?.add(
                IceCandidate(m.sdpMid, (m.sdpMLineIndex ?: 0L).toInt(), m.candidate),
            )
        }
        client.on<ServerMessage.PeerLeft> { m ->
            peers.remove(m.participantId)?.pc?.dispose()
            _ui.update { it.copy(viewerCount = (it.viewerCount - 1).coerceAtLeast(0)) }
        }
        client.on<ServerMessage.SessionEnded> { m ->
            _ui.update { it.copy(ended = endReasonText(m.reason)) }
            stop()
        }
        client.on<ServerMessage.ErrorMessage> { m ->
            _ui.update { it.copy(error = m.userMessage) }
        }
        client.onClosed = {
            if (!stopped) _ui.update { it.copy(ended = "The connection to CrossScreen was lost.") }
        }
    }

    private suspend fun offerTo(participantId: String) {
        val client = signaling ?: return

        val observer = object : PeerObserver() {
            override fun onIceCandidate(candidate: IceCandidate?) {
                candidate ?: return
                client.send(
                    ClientMessage.RtcIce(
                        to = participantId,
                        candidate = candidate.sdp,
                        sdpMid = candidate.sdpMid,
                        sdpMLineIndex = candidate.sdpMLineIndex.toLong(),
                    ),
                )
            }

            override fun onConnectionChange(state: PeerConnection.PeerConnectionState?) {
                _ui.update { ui ->
                    ui.copy(
                        connection = userFacingState(state),
                        viewerCount = if (state == PeerConnection.PeerConnectionState.CONNECTED) {
                            ui.viewerCount.coerceAtLeast(1)
                        } else {
                            ui.viewerCount
                        },
                    )
                }
            }
        }

        val config = PeerConnection.RTCConfiguration(iceServers).apply {
            sdpSemantics = PeerConnection.SdpSemantics.UNIFIED_PLAN
        }
        val pc = WebRtcCore.factory.createPeerConnection(config, observer)
        if (pc == null) {
            _ui.update { it.copy(error = "Could not start the connection.") }
            return
        }
        val ice = IceCandidateQueue().also { it.attach(pc) }
        peers[participantId] = Peer(pc, ice)

        pc.addTrack(track, listOf(STREAM_ID))

        try {
            val offer = pc.createOfferAwait()
            pc.setLocalAwait(offer)
            client.send(ClientMessage.RtcOffer(to = participantId, sdp = offer.description))
        } catch (e: RtcException) {
            Log.w(TAG, "offer failed", e)
            _ui.update { it.copy(error = "Could not start the connection.") }
        }
    }

    private companion object {
        const val TAG = "SharerSession"
        const val STREAM_ID = "screen"
    }
}
